pkg/store/bucket.go

```diff
@@ -2246,18 +2298,29 @@ func (r *bucketIndexReader) ExpandedPostings(ctx context.Context, ms []*labels.M
 		return nil, errors.Wrap(err, "expand")
 	}
 
-	// As of version two all series entries are 16 byte padded. All references
-	// we get have to account for that to get the correct offset.
-	version, err := r.block.indexHeaderReader.IndexVersion()
-	if err != nil {
-		return nil, errors.Wrap(err, "get index version")
-	}
-	if version >= 2 {
-		for i, id := range ps {
-			ps[i] = id * 16
+	// Encode postings to cache. We compress and cache postings before adding
+	// 16 bytes padding in order to make compressed size smaller.
+	dataToCache, compressionDuration, compressionErrors, compressedSize := r.encodePostingsToCache(index.NewListPostings(ps), len(ps))
```

Comment:
I am a little bit concern that we also cache non matched postings to prevent expanding the postings again. But a user could also potentially DDOS the caching layer

Reply:
Is this any different from malicious users just sending a bunch of queries with random matchers which would make the cache full of unused stuff? :thinking: memcached/Redis/etc. support memory/item count limits so it will just expire old items. This problem could probably be solved by using something better than LRU e.g. TinyLFU which doesn't blindly always admit new items into the cache.

Reply:
Redis even has this https://redis.io/docs/reference/eviction/:

```
allkeys-lfu: Keeps frequently used keys; removes least frequently used (LFU) keys
```

Which should solve this problem in theory :smile:

Reply:
Got it. Thanks that makes sense.